### PR TITLE
refactor: plugin system for projects and dependencies

### PR DESCRIPTION
**Pull request by unparalleled-js**
### What I did

fixes: #398

### Checklist

- [ ] Passes all linting checks (pre-commit and CI jobs)
- [ ] New test cases have been added and are passing
- [ ] Documentation has been updated
- [ ] PR title follows [Conventional Commit](https://www.conventionalcommits.org/en/v1.0.0/) standard (will be automatically included in the changelog)

### CODE REVIEW

File: src/ape/managers/config.py [hunk 1]
```diff
@@ -69,7 +82,27 @@ def _plugin_configs(self) -> Dict[str, ConfigItem]:
...
+        if not isinstance(dependencies, list):
+            raise ConfigError("'dependencies' config item must be a list of dicts.")
+
+        decoded_deps = []
+        dep_config_cls: Union[Type[IPFSDependency], Type[GithubDependency]]
+        for dep in dependencies:
+            if "ipfs" in dep:
+                dep_config_cls = IPFSDependency
+            elif "github" in dep:
+                dep_config_cls = GithubDependency
```
Comment (fubuloubu) [inline src/ape/managers/config.py:96]: check out this: https://pydantic-docs.helpmanual.io/usage/model_config/#smart-union

File: src/ape/managers/project.py [hunk 2]
```diff
@@ -59,14 +60,16 @@ def __post_init__(self):
...
-            for n, dep_id in config.dependencies.items()
+            d.name: self._extract_dependency_manifest(d) for d in config.dependencies
         }
 
     def __repr__(self):
         return "<ProjectManager>"
 
-    def _extract_dependency_manifest(self, name: str, download_path: str) -> PackageManifest:
+    def _extract_dependency_manifest(
+        self, dependency_item: Union[IPFSDependency, GithubDependency]
```
Comment (fubuloubu) [inline src/ape/managers/project.py:70]: See? I knew this would come in handy somewhere else!

Maybe our new union type should live in `ape.types` :thinking:

File: src/ape/managers/project.py [hunk 3]
```diff
@@ -86,30 +89,28 @@ def _extract_dependency_manifest(self, name: str, download_path: str) -> Package
...
+            manifest_dict = self._download_manifest(dependency_item, manifest_file_path)
 
         if "name" not in manifest_dict:
             manifest_dict["name"] = name.replace("_", "-")
 
         return PackageManifest(**manifest_dict)
 
-    def _download_manifest(self, name: str, download_path: str, manifest_target_path: Path) -> Dict:
+    def _download_manifest(
+        self, dependency_item: Union[IPFSDependency, GithubDependency], manifest_target_path: Path
```
Comment (fubuloubu) [inline src/ape/managers/project.py:100]: hehe, yes def `Dependency`

File: src/ape/managers/project.py [hunk 4]
```diff
@@ -86,30 +89,28 @@ def _extract_dependency_manifest(self, name: str, download_path: str) -> Package
...
-            except ValueError:
-                raise ValueError("Invalid Github ID. Must be given as <org>/<repo>@<version>")
-
-            # Download manifest
+        elif isinstance(dependency_item, GithubDependency):
             with tempfile.TemporaryDirectory() as temp_dir:
                 temp_project_path = Path(temp_dir) / name
                 temp_project_path.mkdir(exist_ok=True, parents=True)
+                path = dependency_item.github
+                version = dependency_item.version
```
Comment (fubuloubu) [inline src/ape/managers/project.py:113]: Wonder if this could instead be some sort of `abstractmethod` in `DependencyConfig`, and then each dependency config type must implement the same method for fetching and extracting their content to a manfiest file

https://github.com/ApeWorX/ape/pull/431/files#diff-140ab16c725a530ed14f1900c24c98f2f3fc4b9ec4b96a0c439f0daeaf02de26R26

File: src/ape/managers/config.py [hunk 5]
```diff
@@ -19,6 +19,19 @@ class DeploymentConfig(ConfigItem):
...
+    name: str
+
+
+class IPFSDependency(DependencyConfig):
+    ipfs: str
+
+
+class GithubDependency(DependencyConfig):
+    github: str
+    version: str
```
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/managers/config.py:32]: This downloads tar-bars though (releases), in which case the version is required I think.

Would you prefer a `tag` + `clone` approach?

File: src/ape/managers/project.py (hunk 2 above)
Comment (unparalleled-js) [reply src/ape/managers/project.py:70]: Yes, I really do like this typing strategy

File: src/ape/managers/config.py [hunk 6]
```diff
@@ -19,6 +33,65 @@ class DeploymentConfig(ConfigItem):
     contract_type: str
 
 
+class DependencyConfig(ConfigItem):
+    name: str
+    _compilers: "CompilerManager" = None  # type: ignore
```
Comment (unparalleled-js) [inline src/ape/managers/config.py:38]: any dependencies that are not in the PM format require the compilers so they can put that way

(looking at you Github)

Not sure if there is a better way?

File: src/ape/managers/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/managers/config.py:61]: Maybe if they do version, use the release. Otherwise, do clone.

File: docs/userguides/config.md [hunk 7]
```diff
@@ -3,6 +3,36 @@
...
+## Dependencies
+
+Configure dependencies for your ape project.
+Ape downloads and caches dependencies in the `.ape/packages/<name>` directory where `<name>` refers to the name of the dependency.
+
+For example:
+
+```yaml
+dependencies:
+  - name: open_zeppelin
```
Comment (fubuloubu) [inline docs/userguides/config.md:15]: Is this a big deal? lol
```suggestion
  - name: OpenZeppelin
```

File: docs/userguides/config.md [hunk 8]
```diff
@@ -3,6 +3,36 @@
...
+```
+
+will download the [Open Zeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts) package with version `4.4.0`.
+
+To use your dependencies in the `ape-solidity` plugin, configure `import_remappings`:
+
+```yaml
+solidity: 
+  import_remapping:
+    - "@openzeppelin=open_zeppelin"
```
Comment (fubuloubu) [inline docs/userguides/config.md:27]:
```suggestion
    - "@OpenZeppelin=OpenZeppelin"
```

File: docs/userguides/config.md [hunk 9]
```diff
@@ -3,6 +3,36 @@
...
+```
+
+Now, in your solidity files, import `openzeppelin` sources via:
```
Comment (fubuloubu) [inline docs/userguides/config.md:30]:
```suggestion
Now, in your solidity files, import `OpenZeppelin` sources via:
```

File: docs/userguides/config.md [hunk 10]
```diff
@@ -3,6 +3,36 @@
...
+
+```solidity
+import "@openzeppelin/token/ERC721/ERC721.sol";
```
Comment (fubuloubu) [inline docs/userguides/config.md:33]:
```suggestion
import "@OpenZeppelin/token/ERC721/ERC721.sol";
```

**Review by fubuloubu — COMMENTED** [review]
Just me being OCD with CamelCase lol

File: src/ape/managers/config.py [hunk 11]
```diff
@@ -19,6 +33,65 @@ class DeploymentConfig(ConfigItem):
...
+        Returns:
+            PackageManifest
+        """
+
+
+class IPFSDependency(DependencyConfig):
+    ipfs: str
+
+    def extract_manifest(self) -> PackageManifest:
+        response = requests.get(self.ipfs)
```
Comment (fubuloubu) [inline src/ape/managers/config.py:54]: this probably won't work lol... maybe we skip this one for now, and do `GitLab` instead (since it's basically the same)

File: src/ape/managers/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (fubuloubu) [reply src/ape/managers/config.py:61]: yes that sounds right

File: src/ape/managers/config.py [hunk 12]
```diff
@@ -19,6 +33,65 @@ class DeploymentConfig(ConfigItem):
...
+        with tempfile.TemporaryDirectory() as temp_dir:
+            temp_project_path = Path(temp_dir) / self.name
+            temp_project_path.mkdir(exist_ok=True, parents=True)
+            github_client.download_package(self.github, self.version, temp_project_path)
+
+            temp_contracts_path = temp_project_path / "contracts"
+            if not temp_contracts_path.exists():
+                raise ConfigError(
+                    "Dependency does not have a supported file structure. "
+                    "Expecting 'contracts/' path."
```
Comment (fubuloubu) [inline src/ape/managers/config.py:73]: a lot of dapptools/foundry projects will use `src/`

File: src/ape/types/__init__.py [hunk 13]
```diff
@@ -39,7 +41,13 @@
...
 """
 
-AddressType = ChecksumAddress  # type: ignore
+DependencyType = Union["IPFSDependency", "GithubDependency"]
+"""
+A project dependency, such as a :class:`ape.managers.config.IPFSDependency`
+or a :class:`ape.managers.config.GithubDependency`.
+"""
+
+AddressType = NewType("AddressType", HexAddress)  # type: ignore
```
Comment (fubuloubu) [inline src/ape/types/__init__.py:50]: I like using the `ChecksumAddress` assumption though, at least for processing user input, to ensure they didn't make a mistake typing the address

File: docs/userguides/config.md (hunk 7 above)
Comment (unparalleled-js) [reply docs/userguides/config.md:15]: HA! I appreciate a fellow OCD-er TBH.

File: src/ape/types/__init__.py (hunk 13 above)
Comment (unparalleled-js) [reply src/ape/types/__init__.py:50]: `mypy` just really hates this lol I keep fighting it

File: src/ape/managers/config.py (hunk 11 above)
Comment (unparalleled-js) [reply src/ape/managers/config.py:67]: Good call! Let me find an example make the switch

File: src/ape/managers/config.py (hunk 12 above)
Comment (unparalleled-js) [reply src/ape/managers/config.py:104]: Yeah I was wondering this.

This should be configurable on the `DependencyConfig`...
It can mirror our own `contracts_folder` perhaps

File: src/ape/managers/config.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/managers/config.py]: This is done and works great!
Also fixes a bug preventing you from storing multiple versions of a dependecny.

So now:

```yaml
dependencies:
  - name: open_zeppelin
    github: OpenZeppelin/openzeppelin-contracts
```

with clone `main` and still wok and I still have my `4.4.0` as well so I can switch back and forth without having to redownload. Feeling nifty.

File: docs/userguides/config.md [hunk 14]
```diff
@@ -19,6 +19,46 @@ This is useful for projects that compile contracts outside their directory.
...
+```
+
+will download the [Open Zeppelin](https://github.com/OpenZeppelin/openzeppelin-contracts) package with version `4.4.0`.
+
+To use your dependencies in the `ape-solidity` plugin, configure `import_remappings`:
+
+```yaml
+solidity: 
+  import_remapping:
+    - "@penzeppelin=OpenZeppelin/4.4.0"
```
Comment (fubuloubu) [inline docs/userguides/config.md:43]: lulz
```suggestion
    - "@openzeppelin=OpenZeppelin/4.4.0"
```

Comment (fubuloubu) [reply docs/userguides/config.md:43]: p.s. love the version number used there to fight dependency conflicts

File: docs/userguides/config.md [hunk 15]
```diff
@@ -19,6 +19,46 @@ This is useful for projects that compile contracts outside their directory.
...
+```
+
+You can also set the branch and name of the dependency's contract path, e.g.:
+
+```yaml
+dependencies:
+  - name: DappToolsERC20
+    github: dapphub/erc20
+    branch: dappnix
+    contracts_folder: src
```
Comment (fubuloubu) [inline docs/userguides/config.md:59]: :heart_eyes:

File: src/ape/managers/config.py (hunk 11 above)
Comment (fubuloubu) [reply src/ape/managers/config.py:69]: was picturing this would be `ipfs: Qm....` so just the content id. Can probably use a public resolver for now

Comment (fubuloubu) [reply src/ape/managers/config.py:69]: Actually, infura has one too

**Review by fubuloubu — DISMISSED** [review]

File: src/ape/utils.py [hunk 16]
```diff
@@ -368,14 +377,17 @@ def available_plugins(self) -> Set[str]:
...
         """
         Get a release from Github.
 
         Args:
             repo_path (str): The path on Github to the repository,
                             e.g. ``OpenZeppelin/openzeppelin-contracts``.
             version (str): The version of the release to get.
+
+        Returns:
+            GitRelease: The
```
Comment (evanblank3) [inline src/ape/utils.py:390]: The?

File: src/ape/api/projects.py [hunk 17]
```diff
@@ -0,0 +1,50 @@
...
+
+from ape.types import DependencyType
+from ape.utils import abstractdataclass, abstractmethod
+
+if TYPE_CHECKING:
+    from ape.managers.compilers import CompilerManager
+
+
+@abstractdataclass
+class DependencyAPI:
```
Comment (unparalleled-js) [inline src/ape/api/projects.py:13]: This is not the `ProjectAPI` we talked about - this refers to like Github, Gitlab, etc.
The Github implementation is found in `ape.managers.project` currently.

File: src/ape/logging.py [hunk 18]
```diff
@@ -102,6 +102,7 @@ def __init__(self):
         self._logger = _logger
         self._web3_request_manager_logger = _get_logger("web3.RequestManager")
         self._web3_http_provider_logger = _get_logger("web3.providers.HTTPProvider")
+        _logger.setLevel(LogLevel.INFO.value)
```
Comment (unparalleled-js) [inline src/ape/logging.py:105]: fixes race condition

**Comment by unparalleled-js** [comment]
OKAY this PR has reached the point of massive-ness so I am going to open smaller PRs to help with this.

File: src/ape/utils.py [hunk 19]
```diff
@@ -384,6 +396,32 @@ def get_release(self, repo_path: str, version: str):
...
+                            e.g. ``OpenZeppelin/openzeppelin-contracts``.
+            target_path (Path): The local path to store the repo.
+            branch (Optional[str]): The branch to clone. Defaults to the default branch.
+        """
+
+        repo = self.get_repo(repo_path)
+        branch = branch or repo.default_branch
+        logger.info(f"Cloning branch '{branch}' from '{repo.name}'.")
+
+        clone = pygit2.clone_repository(repo.git_url, str(target_path), checkout_branch=branch)
```
Comment (unparalleled-js) [inline src/ape/utils.py:422]: https://github.com/libgit2/pygit2/issues/1119

^ I tried making a progress bar for cloning but was not successful

File: src/ape/managers/project.py [hunk 20]
```diff
@@ -24,11 +25,98 @@
...
 
 from .compilers import CompilerManager
-from .config import ConfigManager
+from .config import ConfigManager, GithubDependencyConfig
 from .converters import ConversionManager
+from .networks import NetworkManager
 
 
-class ProjectManager:
+class BaseProject:
```
Comment (unparalleled-js) [inline src/ape/managers/project.py:33]: So the dependencies and the `ProjectManager` share some things, such as the cache file related logic.
The difference between the two is that the `ProjectManager` has access to the source files whereas the `Dependency` just downloads the source files in a temporary directory and builds the manifest file from that.
Everything related to that is in this file, there a few classes:

`BaseManager`
`ProjectManager`
`DependencyProject` - a base class that any dependency type should be able to use, such as `GitLab`, `IPFS`, etc.
`GithubDependency` - the main implementation we have the downloads or clones from github to build the manifest

Comment (unparalleled-js) [reply src/ape/managers/project.py:33]: I am looking for feedback on the design of this.

I have considered we might want to init an `ape.api.projects` module where we can put the `BaseProject` and the `DependencyProject`.

File: src/ape/managers/compilers.py [hunk 21]
```diff
@@ -78,15 +85,18 @@ def compile(self, contract_filepaths: List[Path]) -> Dict[str, ContractType]:
             Dict[str, ``ContractType``]: A mapping of contract names to their type.
         """
 
+        # NOTE: Must cache 'base_path' before checking registered compilers.
```
Comment (fubuloubu) [inline src/ape/managers/compilers.py:88]: Could do `lrucache` using project_folder?

File: src/ape/managers/compilers.py [hunk 22]
```diff
@@ -95,6 +105,14 @@ def compile(self, contract_filepaths: List[Path]) -> Dict[str, ContractType]:
 
                 contract_types_dict[contract_type.name] = contract_type
 
+        # Fix source IDs for when compiler did not account for base_path,
+        # such as solidity<0.6.9.
```
Comment (fubuloubu) [inline src/ape/managers/compilers.py:109]: Is there a way for this to become contained in the solidity plugin?

File: src/ape/managers/project.py [hunk 23]
```diff
@@ -249,9 +136,108 @@ def sources_missing(self) -> bool:
...
+
+        contracts = self.load_contracts()
+        if attr_name in contracts:
+            contract_type = contracts[attr_name]
+        elif attr_name in self.dependencies:
+            contract_type = self.dependencies[attr_name]  # type: ignore
+        else:
+            # Fixes anomaly when accessing non-ContractType attributes.
+            # Returns normal attribute if exists. Raises 'AttributeError' otherwise.
+            return self.__getattribute__(attr_name)  # type: ignore
```
Comment (fubuloubu) [inline src/ape/managers/project.py:233]: yes, this was the solution @evanblank3

File: src/ape/api/projects.py [hunk 24]
```diff
@@ -0,0 +1,224 @@
...
+if TYPE_CHECKING:
+    from ape.managers.compilers import CompilerManager
+
+
+class ProjectAPI:
+    """
+    An abstract base-class for working with projects. The top-level
+    :class:`ape.managers.project.ProjectManager` implements this class. This class is
+    also a base-class for dependencies (see :class:`~ape.api.projects.DependencyAPI`,
+    such as from Github, that get cached as manifest files.
```
Comment (fubuloubu) [inline src/ape/api/projects.py:29]: Add note about "this can be extended in a plugin to work with non-Ape projects"

File: src/ape/managers/project.py [hunk 25]
```diff
@@ -64,150 +54,47 @@ class ProjectManager:
...
     """
 
-    def __init__(
-        self,
-        path: Path,
-    ) -> None:
-        self.path = Path(path) if isinstance(path, str) else path
-
     def __repr__(self):
         return "<ProjectManager>"
```
Comment (fubuloubu) [inline src/ape/managers/project.py:65]: Move this to `ProjectAPI` as `f"<{self.__class__.__name__}>"`? Maybe add a field about where it came from or what it's name is or whatever.

File: src/ape/managers/compilers.py (hunk 21 above)
Comment (unparalleled-js) [reply src/ape/managers/compilers.py:88]: `project_folder` is not an argument. We just need to grab the value (`self.config.contracts_folder`) before calling `registered_compilers`. Because at that point, if we are compiling a dependency, it will reload the config file and we will lose the path set by the `dependencies:` stanza in our root project config. This is only a problem when the dependency has a different named contracts folder than the default.

File: src/ape/managers/compilers.py (hunk 22 above)
Comment (unparalleled-js) [reply src/ape/managers/compilers.py:109]: yes I found a way

File: src/ape/managers/config.py [hunk 26]
```diff
@@ -83,6 +151,13 @@ def _plugin_configs(self) -> Dict[str, ConfigItem]:
         self.name = user_config.pop("name", "")
         self.version = user_config.pop("version", "")
 
+        dependencies = user_config.pop("dependencies", []) or []
+        if not isinstance(dependencies, list):
+            raise ConfigError("'dependencies' config item must be a list of dicts.")
+
+        # TODO: Handle decoding multiple dependency types
```
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/managers/config.py:158]: This might be part of the work for getting multiple dependency APIs working together? We will have to make `ConfigManager` a `BaseModel`, which I think makes a lot of sense since it is configuration-file driven.

I will create a ticket for this once we are happy with this design, e.g. the new `ProjectAPI` + `DependencyAPI(ProjectAPI)`.

File: src/ape/api/projects.py [hunk 27]
```diff
@@ -0,0 +1,229 @@
...
+    get_all_files_in_directory,
+    get_relative_path,
+    injected_before_use,
+)
+
+if TYPE_CHECKING:
+    from ape.managers.compilers import CompilerManager
+
+
+class ProjectAPI:
```
Comment (unparalleled-js) [inline src/ape/api/projects.py:24]: At this time, I am not sure what to do as far as making this some sort of BaseModel. Thoughts?

Comment (fubuloubu) [reply src/ape/api/projects.py:24]: Seems like `path` would be a good thing to do that for. Right now `path` is a class variable, and this will probably not work too well when you load multiple projects.

File: src/ape/api/projects.py [hunk 28]
```diff
@@ -0,0 +1,229 @@
...
+    @property
+    @abstractmethod
+    def name(self) -> str:
+        """
+        The name of the project.
+        """
+
+    @property
+    @abstractmethod
+    def cache_folder(self) -> Path:
```
Comment (fubuloubu) [inline src/ape/api/projects.py:63]: Does this need to be a part of the API?

File: src/ape/api/projects.py [hunk 29]
```diff
@@ -0,0 +1,229 @@
...
+
+    def __init__(
+        self,
+        path: Path,
+    ) -> None:
+        self.path = Path(path) if isinstance(path, str) else path
+
+    @property
+    @abstractmethod
+    def name(self) -> str:
```
Comment (fubuloubu) [inline src/ape/api/projects.py:56]: This field is in the manifest, wonder if we can just use that

File: src/ape/api/projects.py [hunk 30]
```diff
@@ -0,0 +1,229 @@
...
+    @property
+    @abstractmethod
+    def cache_folder(self) -> Path:
+        """
+        The folder storing the manifest.
+        """
+
+    @property
+    @abstractmethod
+    def version(self) -> str:
```
Comment (fubuloubu) [inline src/ape/api/projects.py:70]: This field is in the manifest, wonder if we can just use that

File: src/ape/api/projects.py [hunk 31]
```diff
@@ -0,0 +1,229 @@
...
+        """
+        The version of the project.
+        """
+
+    @property
+    def manifest_cachefile(self) -> Path:
```
Comment (fubuloubu) [inline src/ape/api/projects.py:76]: This seems sort of like an internal optimization (to eliminate having to reload every time)

File: src/ape/api/projects.py [hunk 32]
```diff
@@ -0,0 +1,229 @@
...
+
+        Returns:
+            pathlib.Path
+        """
+
+        file_name = self.name or "__local__"
+        return self.cache_folder / (file_name + ".json")
+
+    @property
+    def cached_manifest(self) -> Optional[PackageManifest]:
```
Comment (fubuloubu) [inline src/ape/api/projects.py:90]: This seems internal

File: src/ape/api/projects.py [hunk 33]
```diff
@@ -0,0 +1,229 @@
...
+                    hash=compute_checksum(source.read_bytes()),
+                ),
+                urls=[],
+                content=source.read_text(),
+            )
+            for source in contracts_paths
+        }
+
+
+class DependencyAPI(ProjectAPI):
```
Comment (fubuloubu) [inline src/ape/api/projects.py:160]: Was hoping this could be separate from `ProjectAPI`

**Review by fubuloubu — COMMENTED** [review]
Hmm, in general I was thinking about it like this, because these concepts seem orthogonal:

`DependencyAPI`: parse a config item for a dependency and somehow fetch that project

`ProjectAPI`: parse a project and turn it into a manifest somehow (fetching other dependencies, maybe compiling it, etc.)

`ApeProject(ProjectAPI)` would be our native implementation that handles our types of projects (maybe triggered off of `ape-config.yaml` being in the project?)

`GithubDependency(DependencyAPI)` would be supported out of the box too, might be easy to add a few others

Our `ProjectManager` would parse the local project as an `ApeProject` and know how to resolve their dependencies by using `DependencyAPI` methods to fetch and `ProjectAPI` methods to parse them to a manifest, so all of the dependencies are standardized as a set of manifest files for long-term storage between sessions (and eventually deployment, as we want to deploy all manifests upstream of our local project's)

File: src/ape/api/projects.py (hunk 33 above)
Comment (unparalleled-js) [reply src/ape/api/projects.py:160]: If we make them separate, we have to be able to re-create new compilers, config object, etc. Like a whole new ape instance is needed or else nothing works.

Comment (unparalleled-js) [reply src/ape/api/projects.py:160]: this is why I had to make the `compile_external_project` method. Because otherwise, we need to re-initialize a new `CompilerManager` object

File: src/ape/managers/config.py (hunk 1 above)
Comment (unparalleled-js) [reply src/ape/managers/config.py]: Looks like smart union does not support compound types yet, and since this is a `List[DependencyType]` it might not work yet. Looks Pydantic plans to support that eventually though.

File: src/ape/managers/config.py (hunk 26 above)
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/managers/config.py:158]: It looks like pydantic does not support compound types yet for their smart union.

File: src/ape/api/projects.py (hunk 33 above)
Comment (unparalleled-js) [reply src/ape/api/projects.py:160]: Will get this to work the way you want.

**Comment by unparalleled-js** [comment]
Dependencies drawio

**Comment by unparalleled-js** [comment]
Currently, we don't support downloading from private repos on GitHub. Maybe a `ape-github` first class repo would make sense to keep that out of core ape?

**Comment by unparalleled-js** [comment]
@fubuloubu As far as I can tell, it is about ready, though I should probably come back to it with fresh eyes later on.

File: Dockerfile [hunk 34]
```diff
@@ -4,7 +4,7 @@ COPY . .
 # need to use typing-extensions<4 to deal with issue 387 bug: Docker will not build
 # TODO: Figure out a better solution or wait for it to resolve itself.
 RUN pip install typing-extensions==3.10.0.2
-RUN python3 ./setup.py install
+RUN pip install .
```
Comment (fubuloubu) [inline Dockerfile:7]: undo this change, we were playing in `main`

File: src/ape/api/projects.py [hunk 35]
```diff
@@ -0,0 +1,259 @@
...
+        return project_manifest
+
+
+def _load_manifest_from_file(file_path: Path) -> Optional[PackageManifest]:
+    if not file_path.exists():
+        return None
+
+    def handle_corrupt_manifest():
+        logger.warning(f"Existing manifest file '{file_path}' corrupted (re-building).")
+        file_path.unlink()
```
Comment (fubuloubu) [inline src/ape/api/projects.py:243]: I feel like the caller of this should handle deletion, so we don't end up with a weird silent bug

File: src/ape/api/projects.py [hunk 36]
```diff
@@ -0,0 +1,259 @@
...
+
+    except json.JSONDecodeError:
+        handle_corrupt_manifest()
+        return None
+
+    try:
+        return PackageManifest(**manifest_dict)
+    except ValidationError:
+        handle_corrupt_manifest()
+        return None
```
Comment (fubuloubu) [inline src/ape/api/projects.py:259]:
```suggestion
    try:
        manifest_dict = json.loads(file_path.read_text())
        assert isinstance(manifest_dict, dict) and "manifest" in manifest_dict
        return PackageManifest(**manifest_dict)
    except (ValidationError, json.JSONDecodeError, AssertionError):
        logger.warning(f"Existing manifest file '{file_path}' corrupted (re-building).")
        file_path.unlink()
        return None
```

File: src/ape/managers/project.py [hunk 37]
```diff
@@ -4,28 +4,257 @@
...
+    By default, the `:class:`~ape.managers.project.ProjectManager` uses an
+    ``ApeProject`` at the current-working directory.
+    """
+
+    @property
+    def is_valid(self) -> bool:
+        if (self.path / CONFIG_FILE_NAME).exists():
+            return True
+
+        logger.warning(f"'{self.path.name}' is not an 'ApeProject'.")
```
Comment (fubuloubu) [inline src/ape/managers/project.py:38]:
```suggestion
        logger.warning(f"'{self.path.name}' is not an 'ApeProject', but attempting to process as one.")
```

File: src/ape/managers/project.py [hunk 38]
```diff
@@ -4,28 +4,257 @@
...
+        }
+
+        def file_needs_compiling(source: Path) -> bool:
+            path = str(get_relative_path(source, self.contracts_folder))
+
+            if path not in cached_sources:
+                return True  # New file added
+
+            cached = cached_sources[path]
+            cached.compute_checksum(algorithm="md5")
```
Comment (fubuloubu) [inline src/ape/managers/project.py:105]: Hmm, don't really like this (but I know `ethpm-types` would have to be updated)

Feel like it shoudl be `cached.checksum = compute_checksum(cached, algorithm="md5")`

File: src/ape/managers/project.py [hunk 39]
```diff
@@ -4,28 +4,257 @@
...
+
+    plugin_manager: ClassVar[PluginManager] = injected_before_use()  # type: ignore
+    project_manager: ClassVar["ProjectManager"] = injected_before_use()  # type: ignore
+
+    def __init__(self, data_folder: Path):
+        self.DATA_FOLDER = data_folder
+
+    @cached_property
+    def dependency_types(self) -> Dict[str, Type[DependencyAPI]]:
+        dependency_classes = {}
```
Comment (fubuloubu) [inline src/ape/managers/project.py:238]:
```suggestion
        dependency_classes = {
            "github": GithubDependency,
            "local": LocalDependency,
        }
```

File: src/ape/managers/project.py [hunk 40]
```diff
@@ -4,28 +4,257 @@
...
+
+        for _, (config_key, dependency_class) in self.plugin_manager.dependencies:
+            dependency_classes[config_key] = dependency_class
+
+        dependency_classes["github"] = GithubDependency
+        dependency_classes["local"] = LocalDependency
```
Comment (fubuloubu) [inline src/ape/managers/project.py:244]:
```suggestion
```

File: src/ape/managers/project.py [hunk 41]
```diff
@@ -4,28 +4,257 @@
...
+        return dependency_classes
+
+    def decode_dependency(self, config_dependency_data: Dict) -> DependencyAPI:
+        for key, dependency_cls in self.dependency_types.items():
+            if key in config_dependency_data:
+                return dependency_cls(
+                    **config_dependency_data,
+                    project_manager=self.project_manager,
+                    _data_folder=self.DATA_FOLDER,
+                )  # type: ignore
```
Comment (fubuloubu) [inline src/ape/managers/project.py:254]: When pydantic refactor, look into more smartly processing this with `Union` on `ConfigItem`

**Review by fubuloubu — APPROVED** [review]
A lot going on here, but looking really good!

File: src/ape/api/projects.py (hunk 36 above)
Comment (unparalleled-js) [reply src/ape/api/projects.py:259]: `assert` statements is only for debug / testing, they really shouldn't be used in python production code.
If you set `__debug__ = False` (like via `python -O`), the `assert` statements get stripped out of there.

Comment (unparalleled-js) [reply src/ape/api/projects.py:259]: I will try to refactor this to a single except block nonetheless!

File: src/ape/managers/project.py (hunk 41 above)
Comment (unparalleled-js) [reply src/ape/managers/project.py:257]: I have looked into this already struggled due to the fact that it is list of the union type, and pydantic does not yet support that but they are planning on it

File: src/ape/managers/project.py (hunk 38 above)
Comment (unparalleled-js) [reply src/ape/managers/project.py:114]: We can do:

```python
cached.checksum = Checksum(  # type: ignore
    hash=compute_checksum(cached.content.encode("utf8"), algorithm="md5"),
    algorithm="md5",
)
```
without waiting for an update

Comment (unparalleled-js) [reply src/ape/managers/project.py:114]: However, `cachced.content` is optional and `mypy` does not like that

File: src/ape/api/projects.py (hunk 35 above)
Comment (unparalleled-js) [reply src/ape/api/projects.py]: Don

File: src/ape/api/projects.py [hunk 42]
```diff
@@ -0,0 +1,259 @@
...
+
+
+@abstractdataclass
+class ProjectAPI:
+    """
+    An abstract base-class for working with projects.
+    This class can also be extended to a plugin for supporting non-ape projects.
+    """
+
+    compilers: "CompilerManager"
```
[the remark this answers is not in the document]
Comment (unparalleled-js) [reply src/ape/api/projects.py:31]: There is no base API model yet and `injected_before_use()` does not seem to work with `abstractdataclass`.
I am trying to do this but struggling for reasons like this.

Since all the API classes work with way right now, can we do this at the same time as those, as more research is required there.

[the remark this answers is not in the document]
Comment (fubuloubu) [reply src/ape/api/projects.py:31]: for sure that works

File: src/ape/managers/project.py (hunk 38 above)
Comment (fubuloubu) [reply src/ape/managers/project.py:116]: Can work on this later, just seems like a method that should be more pure. The in-place mutation always skeeves me out

**Review by fubuloubu — DISMISSED** [review]

File: src/ape/managers/project.py [hunk 43]
```diff
@@ -4,28 +4,271 @@
...
+            return files
+
+        for extension in self.compilers.registered_compilers:
+            files.extend(self.contracts_folder.rglob(f"*{extension}"))
+
+        return files
+
+    def create_manifest(
+        self, file_paths: Optional[List[Path]] = None, use_cache: bool = True
+    ) -> PackageManifest:
```
Comment (evanblank3) [inline src/ape/managers/project.py:67]: Does this need a doc string?

File: src/ape/managers/project.py [hunk 44]
```diff
@@ -452,85 +553,116 @@ def __getattr__(self, attr_name: str) -> ContractContainer:
...
-        return self.path / "scripts"
+        For example, give it ``HelloWorld`` and it returns
+        ``<absolute-project-path>/contracts/HelloWorld.sol``.
 
-    @property
-    def tests_folder(self) -> Path:
-        """
-        The path to the ``tests/`` directory of the project.
+        Another example is to give it ``contracts/HelloWorld.sol`` and it also
+        returns ``<absolute-project-path>/contracts/HelloWorld.sol``.
```
Comment (evanblank3) [inline src/ape/managers/project.py:596]: Would it be more helpful to change the example here to be a different folder than contracts? This allows for something like `another_contract_folder/HelloWorld.sol` and it would return `<absolute-project-path>/another_contract_folder/HelloWorld.sol`?

**Review by evanblank3 — APPROVED** [review]

File: src/ape/managers/project.py (hunk 43 above)
Comment (unparalleled-js) [reply src/ape/managers/project.py:67]: By not including one, it will automatically used the doc str from the abstractmethod in `ProjectAPI`, which seems desirable. It is not intuitive that is happening but it cuts down on repetitive documentation

File: src/ape/managers/project.py (hunk 44 above)
Comment (unparalleled-js) [reply src/ape/managers/project.py:596]: Yeah, maybe if it said `<contracts-folder>`, it would be more obvious that it is a variable to the project. Good catch!

File: src/ape/managers/project.py (hunk 43 above)
Comment (evanblank3) [reply src/ape/managers/project.py:67]: Missed that this was at the abstract level, thanks!

**Review by evanblank3 — APPROVED** [review]
